amalgamate.mjs

```diff
@@ -0,0 +1,264 @@
+                .map(flag => {
+                    const [name, ...rest] = flag
+                        .slice(definePrefix.length)
+                        .split(defineSep);
+                    let value = rest.join(defineSep);
+                    value = value.replaceAll('\\"', '\"');
+                    return { name, value, flag };
+                });
+            const pathIdentifier = relativePath.replace(/\W+/g, '_');
+            const defines = () => [
+                ...existingDefines(),
+                ...(redefinitions[file] ?? [])
+                    // ...[...new Set(Object.values(redefinitions).flat())]
+                    .map(name => ({ name, value: `${pathIdentifier}__${name}` })),
+            ];
+
+            const flagsWithoutDefines = flags.filter(flag => !flag.startsWith(definePrefix));
+
+            const includeLines = () => [
+                ...defines().map(({ name, value }) => `#define ${name} ${value}`),
+                `#include "${relativePath}"`,
+                ...defines().reverse().map(({ name }) => `#undef ${name}`),
+            ];
+
+            const namespace = relativePath.replace(/\W/g, '_');
+
+            return {
+                directory,
+                file,
+                relativePath,
+                output,
+                command,
+                arguments: args,
+                flags,
+                defines,
+                setCommonDefines,
+                flagsWithoutDefines,
+                includeLines,
+                namespace,
+            };
+        })
+        // Exclude unit tests as that compiles things twice, leading to redefinition errors.
+        .filter(e => !e.defines().find(define => define.name === "_UNIT_TEST_"))
+        .filter(e => !skippedFiles.has(e.relativePath))
+        ;
+
+    function intersection(...sets) {
+        const counter = new Map();
+        for (const set of sets) {
+            for (const e of set) {
+                counter.set(e, (counter.get(e) ?? 0) + 1);
+            }
+        }
+        return [...counter.entries()]
+            .filter(([e, count]) => count == sets.length)
+            .map(([e, count]) => e)
+            ;
+    }
+
+    // defines defined by every translation unit
+    const commonDefines = new Set(intersection(...commands.map(cmd => cmd.defines().map(define => define.flag))));
+
+    const flags = [
+        ...new Set(commands.flatMap(e => e.flagsWithoutDefines)),
+        ...commonDefines,
+        // "-ferror-limit=0",
+        "-Wno-unknown-warning-option",
+        `-I${directory}`,
+    ];
+
+    commands.forEach(cmd => {
+        cmd.setCommonDefines(commonDefines);
+    });
+
+    const includes = [
+        {
+            includeLines: () => [
+                // Need to define `_GNU_SOURCE` the first time we `#include` these.
+                "#define _GNU_SOURCE",
+                "#include <pthread.h>",
+                "#include <sched.h>",
+                "#undef _GNU_SOURCE",
```

Comment:
Do we need to do this?

Reply:
IIRC it didn't build without this, since otherwise the first use of `pthread.h` and `sched.h` didn't have `_GNU_SOURCE` defined yet, so it didn't define the necessary functions, but we can't define `_GNU_SOURCE` for every translation unit, since some have it undefined (I think).